File: frontend/test-runner.js
#!/usr/bin/env node
/**
 * Enhanced test runner for Spicy Todo Frontend
 * Provides comprehensive test execution with detailed reporting
 */

const { spawn, exec } = require('child_process');
const fs = require('fs');
const path = require('path');

class TestRunner {
  constructor(options = {}) {
    this.options = {
      verbose: false,
      coverage: true,
      watch: false,
      updateSnapshot: false,
      ...options
    };
  }

  async runTests(testPattern = '', additionalArgs = []) {
    console.log('🧪 Running Spicy Todo Frontend Tests');
    console.log('=' .repeat(50));

    const args = ['test', '--passWithNoTests'];
    
    // Add coverage if requested
    if (this.options.coverage) {
      args.push('--coverage', '--watchAll=false');
    }
    
    // Add watch mode if requested
    if (this.options.watch) {
      args.push('--watch');
    }
    
    // Add update snapshot if requested
    if (this.options.updateSnapshot) {
      args.push('--updateSnapshot');
    }
    
    // Add test pattern if provided
    if (testPattern) {
      args.push('--testNamePattern', testPattern);
    }
    
    // Add additional arguments
    args.push(...additionalArgs);

    console.log(`Running: npm ${args.join(' ')}`);
    console.log('-'.repeat(50));

    return new Promise((resolve, reject) => {
      const testProcess = spawn('npm', args, {
        stdio: 'inherit',
        shell: true
      });

      testProcess.on('close', (code) => {
        console.log('-'.repeat(50));
        if (code === 0) {
          console.log('✅ All tests passed!');
          this.displayCoverageReport();
        } else {
          console.log(`❌ Tests failed with exit code: ${code}`);
        }
        resolve(code);
      });

      testProcess.on('error', (error) => {
        console.error('❌ Error running tests:', error.message);
        reject(error);
      });
    });
  }

  async runTestSuite(suite) {
    const suites = {
      'unit': '--testPathPattern=__tests__',
      'integration': '--testPathPattern=integration',
      'components': '--testPathPattern=components/__tests__',
      'services': '--testPathPattern=services/__tests__',
      'utils': '--testPathPattern=utils/__tests__',
      'all': ''
    };

    if (!suites[suite]) {
      console.error(`❌ Unknown test suite: ${suite}`);
      console.log(`Available suites: ${Object.keys(suites).join(', ')}`);
      return 1;
    }

    const args = suites[suite] ? [suites[suite]] : [];
    return this.runTests('', args);
  }

  async runSpecificTest(testName) {
    console.log(`🎯 Running specific test: ${testName}`);
    return this.runTests(testName);
  }

  async runTestsWithCoverage() {
    console.log('📊 Running tests with coverage analysis');
    return this.runTests('', ['--coverage', '--watchAll=false']);
  }

  async runTestsInWatchMode() {
    console.log('👀 Running tests in watch mode');
    this.options.watch = true;
    return this.runTests();
  }

  async updateSnapshots() {
    console.log('📸 Updating test snapshots');
    return this.runTests('', ['--updateSnapshot']);
  }

  displayCoverageReport() {
    const coverageDir = path.join(process.cwd(), 'coverage');
    
    if (fs.existsSync(coverageDir)) {
      const lcovPath = path.join(coverageDir, 'lcov-report', 'index.html');
      
      if (fs.existsSync(lcovPath)) {
        console.log('\n📈 Coverage Report Generated:');
        console.log(`   HTML Report: file://${lcovPath}`);
        
        // Try to read coverage summary
        const summaryPath = path.join(coverageDir, 'coverage-summary.json');
        if (fs.existsSync(summaryPath)) {
          try {
            const summary = JSON.parse(fs.readFileSync(summaryPath, 'utf8'));
            const totals = summary.total;
            
            console.log('\n📊 Coverage Summary:');
            console.log(`   Lines: ${totals.lines.pct}% (${totals.lines.covered}/${totals.lines.total})`);
            console.log(`   Functions: ${totals.functions.pct}% (${totals.functions.covered}/${totals.functions.total})`);
            console.log(`   Branches: ${totals.branches.pct}% (${totals.branches.covered}/${totals.branches.total})`);
            console.log(`   Statements: ${totals.statements.pct}% (${totals.statements.covered}/${totals.statements.total})`);
          } catch (error) {
            console.log('   Could not parse coverage summary');
          }
        }
      }
    }
  }

  async checkTestEnvironment() {
    console.log('🔍 Checking test environment...');
    
    const checks = [
      {
        name: 'Node.js version',
        check: () => {
          const version = process.version;
          const major = parseInt(version.slice(1).split('.')[0]);
          return { success: major >= 14, message: `Node.js ${version}` };
        }
      },
      {
        name: 'npm package.json',
        check: () => {
          const packagePath = path.join(process.cwd(), 'package.json');
          return { 
            success: fs.existsSync(packagePath), 
            message: fs.existsSync(packagePath) ? 'Found package.json' : 'package.json not found' 
          };
        }
      },
      {
        name: 'Jest configuration',
        check: () => {
          const jestConfig = path.join(process.cwd(), 'jest.config.js');
          return { 
            success: fs.existsSync(jestConfig), 
            message: fs.existsSync(jestConfig) ? 'Found jest.config.js' : 'jest.config.js not found' 
          };
        }
      },
      {
        name: 'Test dependencies',
        check: async () => {
          return new Promise((resolve) => {
            exec('npm list @testing-library/react @testing-library/jest-dom', (error, stdout) => {
              resolve({
                success: !error,
                message: error ? 'Testing dependencies missing' : 'Testing dependencies installed'
              });
            });
          });
        }
      },
      {
        name: 'Test files',
        check: () => {
          const testDirs = [
            'src/__tests__',
            'src/components/__tests__',
            'src/services/__tests__',
            'src/utils/__tests__'
          ];
          
          const existingDirs = testDirs.filter(dir => fs.existsSync(path.join(process.cwd(), dir)));
          return {
            success: existingDirs.length > 0,
            message: `${existingDirs.length}/${testDirs.length} test directories found`
          };
        }
      }
    ];

    let allPassed = true;
    
    for (const check of checks) {
      try {
        const result = await check.check();
        console.log(`   ${result.success ? '✅' : '❌'} ${check.name}: ${result.message}`);
        if (!result.success) allPassed = false;
      } catch (error) {
        console.log(`   ❌ ${check.name}: Error - ${error.message}`);
        allPassed = false;
      }
    }

    if (allPassed) {
      console.log('\n✅ Test environment is properly set up');
    } else {
      console.log('\n❌ Test environment has issues');
    }

    return allPassed;
  }

  async generateTestReport() {
    console.log('📋 Generating comprehensive test report...');
    
    const report = {
      timestamp: new Date().toISOString(),
      environment: {
        nodeVersion: process.version,
        platform: process.platform,
        arch: process.arch
      },
      testSuites: {}
    };

    // Run different test suites and collect results
    const suites = ['components', 'services', 'utils', 'integration'];
    
    for (const suite of suites) {
      console.log(`\n🧪 Running ${suite} tests...`);
      const exitCode = await this.runTestSuite(suite);
      report.testSuites[suite] = {
        passed: exitCode === 0,
        exitCode
      };
    }

    // Generate coverage report
    console.log('\n📊 Generating coverage report...');
    await this.runTestsWithCoverage();

    // Save report
    const reportPath = path.join(process.cwd(), 'test-report.json');
    fs.writeFileSync(reportPath, JSON.stringify(report, null, 2));
    
    console.log(`\n📄 Test report saved to: ${reportPath}`);
    
    return report;
  }
}

// CLI interface
async function main() {
  const args = process.argv.slice(2);
  const runner = new TestRunner();

  if (args.length === 0) {
    // Default: run all tests with coverage
    return await runner.runTestsWithCoverage();
  }

  const command = args[0];

  switch (command) {
    case '--suite':
      const suite = args[1];
      if (!suite) {
        console.error('❌ Please specify a test suite');
        console.log('Available suites: unit, integration, components, services, utils, all');
        return 1;
      }
      return await runner.runTestSuite(suite);

    case '--test':
      const testName = args[1];
      if (!testName) {
        console.error('❌ Please specify a test name pattern');
        return 1;
      }
      return await runner.runSpecificTest(testName);

    case '--watch':
      return await runner.runTestsInWatchMode();

    case '--update-snapshots':
      return await runner.updateSnapshots();

    case '--coverage':
      return await runner.runTestsWithCoverage();

    case '--check-env':
      const envOk = await runner.checkTestEnvironment();
      return envOk ? 0 : 1;

    case '--report':
      await runner.generateTestReport();
      return 0;

    case '--help':
      console.log(`
🧪 Spicy Todo Frontend Test Runner

Usage: node test-runner.js [command] [options]

Commands:
  (no args)           Run all tests with coverage
  --suite <name>      Run specific test suite (unit, integration, components, services, utils, all)
  --test <pattern>    Run tests matching pattern
  --watch             Run tests in watch mode
  --update-snapshots  Update test snapshots
  --coverage          Run tests with coverage analysis
  --check-env         Check test environment setup
  --report            Generate comprehensive test report
  --help              Show this help message

Examples:
  node test-runner.js --suite components
  node test-runner.js --test "TodoForm"
  node test-runner.js --watch
  node test-runner.js --report
      `);
      return 0;

    default:
      console.error(`❌ Unknown command: ${command}`);
      console.log('Run "node test-runner.js --help" for usage information');
      return 1;
  }
}

// Run if called directly
if (require.main === module) {
  main().then(code => process.exit(code));
}

module.exports = TestRunner;

File: api/pyfast-api/tests/test_runner.py
#!/usr/bin/env python3
"""
Enhanced test runner for Spicy Todo API
Provides comprehensive test execution with detailed reporting
"""

import sys
import os
import subprocess
import argparse
import time
from pathlib import Path

def run_tests(
    test_path=None,
    verbose=False,
    coverage=True,
    html_report=True,
    parallel=False,
    markers=None,
    fail_fast=False,
    performance=False,
    security=False,
    integration=False
):
    """
    Run tests with various options and configurations
    """
    
    # Base pytest command
    cmd = ['python', '-m', 'pytest']
    
    # Add test path or default to tests directory
    if test_path:
        cmd.append(test_path)
    else:
        cmd.append('tests/')
    
    # Add verbosity
    if verbose:
        cmd.append('-v')
    else:
        cmd.append('-q')
    
    # Add coverage
    if coverage:
        cmd.extend([
            '--cov=.',
            '--cov-report=term-missing',
            '--cov-report=xml:coverage.xml'
        ])
        
        if html_report:
            cmd.append('--cov-report=html:htmlcov')
    
    # Add parallel execution
    if parallel:
        cmd.extend(['-n', 'auto'])
    
    # Add markers
    if markers:
        cmd.extend(['-m', markers])
    
    # Add fail fast
    if fail_fast:
        cmd.append('-x')
    
    # Add specific test categories
    test_files = []
    if performance:
        test_files.append('tests/test_performance.py')
    if security:
        test_files.append('tests/test_security.py')
    if integration:
        test_files.append('tests/test_e2e_workflows.py')
    
    if test_files:
        cmd.extend(test_files)
    
    # Add additional options
    cmd.extend([
        '--tb=short',
        '--strict-markers',
        '--disable-warnings'
    ])
    
    print(f"Running command: {' '.join(cmd)}")
    print("-" * 60)
    
    # Run the tests
    start_time = time.time()
    result = subprocess.run(cmd, capture_output=False)
    end_time = time.time()
    
    print("-" * 60)
    print(f"Test execution completed in {end_time - start_time:.2f} seconds")
    
    return result.returncode

def run_specific_test_suite(suite_name):
    """Run specific test suites"""
    
    test_suites = {
        'unit': 'tests/test_*.py -k "not integration and not performance and not security"',
        'api': 'tests/test_api_endpoints.py',
        'models': 'tests/test_models.py',
        'database': 'tests/test_database.py',
        'middleware': 'tests/test_middleware.py',
        'logging': 'tests/test_logging_config.py',
        'performance': 'tests/test_performance.py',
        'security': 'tests/test_security.py',
        'integration': 'tests/test_e2e_workflows.py',
        'all': None
    }
    
    if suite_name not in test_suites:
        print(f"Unknown test suite: {suite_name}")
        print(f"Available suites: {', '.join(test_suites.keys())}")
        return 1
    
    if suite_name == 'all':
        return run_tests(verbose=True, coverage=True)
    else:
        return run_tests(test_path=test_suites[suite_name], verbose=True)

def check_test_environment():
    """Check if test environment is properly set up"""
    
    print("Checking test environment...")
    
    # Check if pytest is installed
    try:
        import pytest
        print(f"✓ pytest {pytest.__version__} is installed")
    except ImportError:
        print("✗ pytest is not installed")
        return False
    
    # Check if coverage is available
    try:
        import coverage
        print(f"✓ coverage {coverage.__version__} is available")
    except ImportError:
        print("✗ coverage is not installed")
        return False
    
    # Check if test files exist
    test_files = [
        'tests/test_api_endpoints.py',
        'tests/test_models.py',
        'tests/test_database.py',
        'tests/test_middleware.py',
        'tests/test_logging_config.py',
        'tests/test_performance.py',
        'tests/test_security.py',
        'tests/test_e2e_workflows.py'
    ]
    
    missing_files = []
    for test_file in test_files:
        if Path(test_file).exists():
            print(f"✓ {test_file} exists")
        else:
            print(f"✗ {test_file} is missing")
            missing_files.append(test_file)
    
    if missing_files:
        print(f"\nMissing test files: {missing_files}")
        return False
    
    print("\n✓ Test environment is properly set up")
    return True

def generate_test_report():
    """Generate comprehensive test report"""
    
    print("Generating test report...")
    
    # Run tests with coverage
    result = run_tests(verbose=True, coverage=True, html_report=True)
    
    if result == 0:
        print("\n✓ All tests passed!")
        
        # Check if coverage report was generated
        if Path('htmlcov/index.html').exists():
            print("✓ HTML coverage report generated: htmlcov/index.html")
        
        if Path('coverage.xml').exists():
            print("✓ XML coverage report generated: coverage.xml")
    else:
        print(f"\n✗ Tests failed with exit code: {result}")
    
    return result

def main():
    """Main entry point"""
    
    parser = argparse.ArgumentParser(description='Spicy Todo API Test Runner')
    parser.add_argument('--suite', choices=[
        'unit', 'api', 'models', 'database', 'middleware', 
        'logging', 'performance', 'security', 'integration', 'all'
    ], help='Run specific test suite')
    parser.add_argument('--check-env', action='store_true', help='Check test environment')
    parser.add_argument('--report', action='store_true', help='Generate comprehensive test report')
    parser.add_argument('--verbose', '-v', action='store_true', help='Verbose output')
    parser.add_argument('--no-coverage', action='store_true', help='Disable coverage reporting')
    parser.add_argument('--parallel', '-p', action='store_true', help='Run tests in parallel')
    parser.add_argument('--fail-fast', '-x', action='store_true', help='Stop on first failure')
    parser.add_argument('--performance', action='store_true', help='Run performance tests')
    parser.add_argument('--security', action='store_true', help='Run security tests')
    parser.add_argument('--integration', action='store_true', help='Run integration tests')
    parser.add_argument('--markers', '-m', help='Run tests with specific markers')
    
    args = parser.parse_args()
    
    # Check environment if requested
    if args.check_env:
        if not check_test_environment():
            return 1
    
    # Generate report if requested
    if args.report:
        return generate_test_report()
    
    # Run specific test suite
    if args.suite:
        return run_specific_test_suite(args.suite)
    
    # Run tests with specified options
    return run_tests(
        verbose=args.verbose,
        coverage=not args.no_coverage,
        parallel=args.parallel,
        fail_fast=args.fail_fast,
        performance=args.performance,
        security=args.security,
        integration=args.integration,
        markers=args.markers
    )

if __name__ == '__main__':
    sys.exit(main())
